internals/cli/cmd_add-identities.go

```diff
@@ -0,0 +1,103 @@
+// Copyright (c) 2024 Canonical Ltd
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package cli
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"strconv"
+
+	"github.com/canonical/go-flags"
+	"gopkg.in/yaml.v3"
+
+	"github.com/canonical/pebble/client"
+)
+
+const cmdAddIdentitiesSummary = "Add new identities"
+const cmdAddIdentitiesDescription = `
+The add-identities command adds one or more new identities.
+
+The named identities must not yet exist.
+
+For example, to add a local admin named "bob", use YAML like this:
+
+> identities:
+>     bob:
+>         access: admin
+>         local:
+>             user-id: 42
+`
+
+type cmdAddIdentities struct {
+	client *client.Client
+
+	From string `long:"from" required:"1"`
+}
+
+func init() {
+	AddCommand(&CmdInfo{
+		Name:        "add-identities",
+		Summary:     cmdAddIdentitiesSummary,
+		Description: cmdAddIdentitiesDescription,
+		ArgsHelp: map[string]string{
+			"--from": "Path of YAML file to read identities from (required)",
+		},
+		New: func(opts *CmdOptions) flags.Commander {
+			return &cmdAddIdentities{client: opts.Client}
+		},
+	})
+}
+
+func (cmd *cmdAddIdentities) Execute(args []string) error {
+	if len(args) > 0 {
+		return ErrExtraArgs
+	}
+
+	identities, err := readIdentities(cmd.From)
+	if err != nil {
+		return err
+	}
+	err = cmd.client.AddIdentities(identities)
```

Comment:
The only protection from nil identities in the map read from the file is in the state validate method for the entity. We probably should do some client side validation to return nice errors.

Reply:
I somewhat disagree that all the validation should be on the client side (we need to do it on the server side anyway). However, I've added a bit of error handling and made things a bit more consistent on the server side.